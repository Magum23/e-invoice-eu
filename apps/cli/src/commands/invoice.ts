// FIXME! Name all the commands SubjectCommand so that we can import the
// Invoice interface in a normal fashion.
import { Invoice as CoreInvoice, InvoiceFile, InvoiceService, InvoiceServiceOptions, MappingService } from '@e-invoice-eu/core';
import { Textdomain } from '@esgettext/runtime';
import * as fs from 'fs/promises';
import { lookup } from 'mime-types';
import * as path from 'path';
import yargs, { InferredOptionTypes } from 'yargs';

import { Command } from '../command';
import { coerceOptions, OptSpec } from '../optspec';
import { Package } from '../package';
import { safeStdoutBufferWrite, safeStdoutWrite } from '../safe-stdout-write';

const gtx = Textdomain.getInstance('e-invoice-eu-cli');

const options: {
	format: OptSpec;
	output: OptSpec;
	invoice: OptSpec;
	mapping: OptSpec;
	data: OptSpec;
	lang: OptSpec;
	pdf: OptSpec;
	'pdf-id': OptSpec;
	'pdf-description': OptSpec,
	attachment: OptSpec;
	'attachment-id': OptSpec;
	'attachment-description': OptSpec;
	'attachment-mimetype': OptSpec;
	'embed-pdf': OptSpec;
} = {
	format: {
		group: gtx._('Format selection'),
		alias: ['f'],
		type: 'string',
		demandOption: true,
		describe: gtx._("invoice format (case-insensitive), try 'format --list' for a list of allowed values"),
	},
	output: {
		group: gtx._('Output file location'),
		alias: ['o'],
		type: 'string',
		demandOption: false,
		describe: gtx._('write output to specified file instead of standard output'),
	},
	invoice: {
		group: gtx._('Input data'),
		alias: ['i'],
		type: 'string',
		conflicts: ['mapping'],
		demandOption: false,
		describe: gtx._('invoice data as JSON'),
	},
	mapping: {
		group: gtx._('Input data'),
		alias: ['m'],
		type: 'string',
		conflicts: ['invoice'],
		demandOption: false,
		describe: gtx._('mapping file (YAML or JSON), mandatory for spreadsheet data input'),
	},
	data: {
		group: gtx._('Input data'),
		alias: ['d'],
		type: 'string',
		demandOption: false,
		describe: gtx._('invoice spreadsheet data, mandatory for spreadsheet data input'),
	},
	pdf: {
		group: gtx._('Input data'),
		alias: ['p'],
		type: 'string',
		demandOption: false,
		describe: gtx._('PDF version of the invoice'),
	},
	'pdf-id': {
		group: gtx._('Input data'),
		type: 'string',
		demandOption: false,
		describe: gtx._('ID of the embedded PDF, defaults to the document number')
	},
	'pdf-description': {
		group: gtx._('Input data'),
		type: 'string',
		demandOption: false,
		describe: gtx._('optional description of the embedded PDF')
	},
	attachment: {
		group: gtx._('Input data'),
		alias: ['a'],
		type: 'string',
		multi: true,
		demandOption: false,
		describe: gtx._('arbitrary number of attachments'),
	},
	'attachment-id': {
		group: gtx._('Input data'),
		type: 'string',
		multi: true,
		demandOption: false,
		describe: gtx._('optional ids of the attachments')
	},
	'attachment-description': {
		group: gtx._('Input data'),
		type: 'string',
		multi: true,
		demandOption: false,
		describe: gtx._('optional descriptions of the attachments')
	},
	'attachment-mimetype': {
		group: gtx._('Input data'),
		alias: ['attachment-mime-type'],
		type: 'string',
		multi: true,
		demandOption: false,
		describe: gtx._('optional MIME types of the attachments')
	},
	lang: {
		group: gtx._('Invoice details'),
		alias: ['l'],
		type: 'string',
		demandOption: false,
		default: 'en',
		describe: gtx._('invoice language code'),
	},
	'embed-pdf': {
		group: gtx._('Invoice details'),
		type: 'boolean',
		demandOption: false,
		describe: gtx._('embed a PDF version of the invoice'),
	},
};

export type ConfigOptions = InferredOptionTypes<typeof options>;

export class Invoice implements Command {
	description(): string {
		return gtx._('Create an e-invoice from spreadsheet data or JSON.');
	}

	aliases(): Array<string> {
		return [];
	}

	build(argv: yargs.Argv): yargs.Argv<object> {
		return argv.options(options);
	}

	private async addPdf(options: InvoiceServiceOptions, configOptions: ConfigOptions) {
		if (configOptions.pdf) {
			const invoiceFile: InvoiceFile = {
				buffer: await fs.readFile(configOptions.pdf as string),
				filename: path.basename(configOptions.pdf as string),
				mimetype: 'application/pdf',
			};

			options.pdf = invoiceFile;
		}
	}

	private checkConfigOptions(configOptions: ConfigOptions) {
		if (typeof configOptions.invoice === 'undefined' && typeof configOptions.mapping === 'undefined') {
			throw new Error(gtx._("One of the options '--invoice' or '--mapping' is mandatory."));
		} else if (typeof configOptions.mapping !== 'undefined'
			&& typeof configOptions.data === 'undefined') {
			throw new Error(gtx._('No invoice spreadsheet specified.'));
		}
	}

	private async addAttachments(options: InvoiceServiceOptions, configOptions: ConfigOptions) {
		if (!configOptions.attachment) return;

		const attachments = configOptions.attachment as string[];
		for (let i = 0; i < attachments.length; ++i) {
			const filename = attachments[i];
			const basename = path.basename(filename);
			const mimetype = configOptions['attachment-mimetype']?.[i] ?? lookup(basename);

			if (!mimetype) {
				throw new Error(gtx._x("cannot guess MIME type of attachment '{filename}'!", { filename }));
			}

			const file: InvoiceFile = {
				buffer: await fs.readFile(filename),
				filename: basename,
				mimetype,
			};

			options.attachments.push({
				file,
				id: configOptions['attachment-id']?.[i],
				description: configOptions['attachment-description']?.[i],
			});
		}
	}

	private async createInvoice(options: InvoiceServiceOptions, configOptions: ConfigOptions): Promise<string | Buffer> {
		let invoiceData: CoreInvoice;

		const format = configOptions.format as string;

		if (typeof configOptions.invoice !== 'undefined') {
			const filename = configOptions.invoice as string;

			const json = await fs.readFile(filename, 'utf-8');
			try {
				invoiceData = JSON.parse(json) as CoreInvoice;
			} catch(e) {
				throw new Error(`${filename}: ${e.message}`);
			}
		} else {
			const mapping = await fs.readFile((configOptions.mapping as string), 'utf-8');
			const data = await fs.readFile((configOptions.data as string));
			const mappingService = new MappingService(console);
			invoiceData = mappingService.transform(
				format.toLowerCase(),
				mapping,
				data,
			);
		}

		const invoiceService = new InvoiceService(console);

		return await invoiceService.generate(invoiceData, options);
	}

	private async doRun(configOptions: ConfigOptions) {
		console.dir(configOptions);
		const options: InvoiceServiceOptions = {
			format: configOptions.format as string,
			lang: configOptions.lang as string,
			attachments: [],
		};

		this.checkConfigOptions(configOptions);
		await this.addPdf(options, configOptions);
		await this.addAttachments(options, configOptions);

		const document = await this.createInvoice(options, configOptions);
		if (typeof document === 'string') {
			if (typeof configOptions.output === 'undefined') {
				safeStdoutWrite(document);
			} else {
				await fs.writeFile(configOptions.output as string, document, 'utf-8');
			}
		} else {
			if (typeof configOptions.output === 'undefined') {
				safeStdoutBufferWrite(document);
			}
		}
	}

	public async run(argv: yargs.Arguments): Promise<number> {
		const configOptions = argv as unknown as ConfigOptions;

		if (!coerceOptions(argv, options)) {
			return 1;
		}

		try {
			await this.doRun(configOptions);
			return 0;
		} catch (e) {
			console.error(
				gtx._x('{programName}: {error}', {
					programName: Package.getName(),
					error: e,
				}),
			);

			return 1;
		}
	}
}
