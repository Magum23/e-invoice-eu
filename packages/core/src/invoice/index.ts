/* istanbul ignore file */
export { Invoice } from './invoice.interface';
export * from './invoice.schema';
export * from './invoice.service';
