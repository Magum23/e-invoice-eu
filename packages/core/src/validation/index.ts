/* istanbul ignore file */
export * from './validation.service';
