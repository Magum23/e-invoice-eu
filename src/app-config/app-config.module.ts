import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';

import { AppConfigService } from './app-config.service';

@Module({
	imports: [
		ConfigModule.forRoot({
			isGlobal: true, // Ensures ConfigModule is available globally
		}),
	],
	providers: [AppConfigService],
	exports: [AppConfigService], // Export AppConfigService so it can be used in other modules
})
export class AppConfigModule {}
