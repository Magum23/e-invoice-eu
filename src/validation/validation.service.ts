import { Injectable, Logger } from '@nestjs/common';
import { ErrorObject, ValidateFunction, ValidationError } from 'ajv';

@Injectable()
export class ValidationService {
	private readonly logger = new Logger(ValidationService.name);

	validate<T>(id: string, f: ValidateFunction<T>, data: unknown): T {
		if (f(data)) {
			return data as T;
		} else {
			this.logger.error(`Invalid ${id}: ${JSON.stringify(f.errors, null, 4)}`);
			throw new ValidationError(f.errors as ErrorObject[]);
		}
	}
}
