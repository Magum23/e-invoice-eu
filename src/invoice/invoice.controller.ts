import {
	BadRequestException,
	Body,
	Controller,
	HttpStatus,
	InternalServerErrorException,
	Logger,
	Param,
	Post,
	Res,
	UploadedFiles,
	UseInterceptors,
} from '@nestjs/common';
import { FileFieldsInterceptor } from '@nestjs/platform-express';
import {
	ApiBody,
	ApiConsumes,
	ApiParam,
	ApiResponse,
	ApiTags,
} from '@nestjs/swagger';
import { ValidationError } from 'ajv/dist/2019';
import { Response } from 'express';

import { InvoiceService } from './invoice.service';
import { MappingService } from '../mapping/mapping.service';

@ApiTags('invoice')
@Controller('invoice')
export class InvoiceController {
	constructor(
		private readonly invoiceService: InvoiceService,
		private readonly mappingService: MappingService,
		private readonly logger: Logger,
	) {}

	@Post('transform-and-create/:format')
	@ApiParam({
		name: 'format',
		type: String,
		description:
			'The format of the invoice to transform and create, for' +
			' example "UBL" or "CII". Case does not matter.',
		example: 'UBL',
	})
	@ApiConsumes('multipart/form-data')
	@ApiBody({
		required: true,
		schema: {
			type: 'object',
			properties: {
				data: {
					type: 'string',
					format: 'binary',
					description: 'The spreadsheet to be transformed.',
				},
				mapping: {
					type: 'string',
					format: 'binary',
					description:
						'The mapping from spreadsheet data to the internal format as YAML or JSON.',
				},
				pdf: {
					type: 'string',
					format: 'binary',
					nullable: true,
					description:
						'Optional PDF version of the invoice.  For Factur-X/ZUGFeRD, if no PDF is uploaded, one is generated from the Spreadsheet with the help of LibreOffice.',
				},
				attachment: {
					type: 'array',
					nullable: true,
					description:
						'An arbitrary number of supplementary attachments. **Not yet implemented!**',
					items: {
						type: 'string',
						format: 'binary',
						description: 'The individual attachment',
					},
				},
				description: {
					type: 'array',
					nullable: true,
					description:
						'Optional descriptions for each supplementary attachment',
					items: {
						type: 'string',
						description: 'Description for the corresponding attachment',
					},
				},
			},
		},
	})
	@ApiResponse({
		status: 201,
		description:
			'Creation successful. The output is an invoice' +
			' document as either XML or PDF.',
	})
	@ApiResponse({
		status: 400,
		description: 'Bad request with error details',
	})
	@UseInterceptors(
		FileFieldsInterceptor([
			{ name: 'data', maxCount: 1 },
			{ name: 'mapping', maxCount: 1 },
			{ name: 'pdf', maxCount: 1 },
			{ name: 'attachment' }, // FIXME! How to set maxCount asynchronously?
		]),
	)
	async transformAndCreate(
		@Res() response: Response,
		@Param('format') format: string,
		@UploadedFiles()
		files: {
			data?: Express.Multer.File[];
			mapping?: Express.Multer.File[];
			pdf?: Express.Multer.File[];
			attachment?: Express.Multer.File[];
		},

		@Body() body: { description?: string[]; mimeType?: string[] },
	) {
		const { data, mapping, pdf } = files;

		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		const descriptions = body.description || [];
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		const mimeTypes = body.mimeType || [];

		if (!data) {
			throw new BadRequestException('No invoice file uploaded');
		}

		if (!mapping) {
			throw new BadRequestException('No mapping file uploaded');
		}

		try {
			const invoice = this.mappingService.transform(
				format.toLowerCase(),
				mapping[0].buffer.toString(),
				data[0].buffer,
			);

			const document = await this.invoiceService.generate(invoice, {
				format: format.toLowerCase(),
				data: data[0].buffer,
				dataName: data ? data[0].originalname : undefined,
				pdf: pdf ? pdf[0].buffer : undefined,
			});
			if (typeof document === 'string') {
				response.set('Content-Type', 'application/xml');
			} else {
				response.set('Content-Type', 'application/pdf');
			}
			response.status(HttpStatus.CREATED).send(document);
		} catch (error) {
			if (error instanceof ValidationError) {
				throw new BadRequestException({
					message: 'Transformation failed.',
					details: error,
				});
			} else {
				this.logger.error(`unknown error: ${error.message}\n${error.stack}`);

				throw new InternalServerErrorException();
			}
		}
	}
}
