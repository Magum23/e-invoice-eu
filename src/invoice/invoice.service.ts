import { Injectable } from '@nestjs/common';

import { Invoice } from './invoice.interface';
import { FormatFactoryService } from '../format/format.factory.service';

@Injectable()
export class InvoiceService {
	constructor(private readonly formatFactoryService: FormatFactoryService) {}

	generate(format: string, invoice: Invoice): string {
		const formatter = this.formatFactoryService.createFormatService(format);
		const xml = formatter.generate(invoice);

		return xml;
	}
}
