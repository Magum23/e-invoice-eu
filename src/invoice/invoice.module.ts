/* istanbul ignore file */
import { Logger, Module } from '@nestjs/common';

import { InvoiceController } from './invoice.controller';
import { InvoiceService } from './invoice.service';
import { FormatFactoryService } from '../format/format.factory.service';
import { FormatModule } from '../format/format.module';
import { MappingModule } from '../mapping/mapping.module';
import { MappingService } from '../mapping/mapping.service';
import { SerializerModule } from '../serializer/serializer.module';
import { SerializerService } from '../serializer/serializer.service';
import { ValidationModule } from '../validation/validation.module';
import { ValidationService } from '../validation/validation.service';

@Module({
	imports: [FormatModule, MappingModule, SerializerModule, ValidationModule],
	controllers: [InvoiceController],
	providers: [
		Logger,
		FormatFactoryService,
		InvoiceService,
		MappingService,
		SerializerService,
		ValidationService,
	],
	exports: [InvoiceService],
})
export class InvoiceModule {}
