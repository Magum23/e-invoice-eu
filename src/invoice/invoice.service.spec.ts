import { Test, TestingModule } from '@nestjs/testing';
import { ErrorObject, ValidationError } from 'ajv/dist/2019';

import { InvoiceService } from './invoice.service';
import { AppConfigService } from '../app-config/app-config.service';
import { FormatFactoryService } from '../format/format.factory.service';
import { ValidationService } from '../validation/validation.service';

describe('InvoiceService', () => {
	let service: InvoiceService;
	let validationService: ValidationService;

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				InvoiceService,
				AppConfigService,
				{
					provide: 'AppConfigService',
					useValue: {},
				},
				FormatFactoryService,
				{
					provide: 'FormatFactoryService',
					useValue: {},
				},
				ValidationService,
				{
					provide: 'ValidationService',
					useValue: {
						validate: jest.fn(),
					},
				},
			],
		}).compile();

		service = module.get<InvoiceService>(InvoiceService);
		validationService = module.get<ValidationService>(ValidationService);
	});

	it('should be defined', () => {
		expect(service).toBeDefined();
	});

	it('should validate input data and create an invoice', async () => {
		const validateSpy = jest
			.spyOn(validationService, 'validate')
			.mockReturnValue({ 'ubl:Invoice': {} });

		const input = {} as unknown;
		const got = await service.generate(input, {
			lang: 'en-us',
			format: 'UBL',
			attachments: [],
		});

		expect(got).toMatchSnapshot();

		expect(validateSpy).toHaveBeenCalledTimes(1);
		expect(validateSpy).toHaveBeenCalledWith(
			'invoice data',
			expect.anything(),
			input,
		);

		validateSpy.mockRestore();
	});

	it('should reject invalid input data', async () => {
		const validateSpy = jest
			.spyOn(validationService, 'validate')
			.mockImplementationOnce(() => {
				// FIXME: Use more specific error here!
				throw new ValidationError([] as ErrorObject[]);
			});

		const input = {} as unknown;

		try {
			await service.generate(input, {
				lang: 'en-us',
				format: 'UBL',
				attachments: [],
			});
			fail('Expected generate() to throw a ValidationError.');
		} catch (error) {
			expect(error).toBeInstanceOf(ValidationError);
			expect(error.errors).toEqual([]);
		}

		expect(validateSpy).toHaveBeenCalledTimes(1);
		expect(validateSpy).toHaveBeenCalledWith(
			'invoice data',
			expect.anything(),
			input,
		);

		validateSpy.mockRestore();
	});
});
