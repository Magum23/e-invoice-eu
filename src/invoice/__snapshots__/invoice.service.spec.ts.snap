// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceService should create an invoice 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<Invoice xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2" xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2" xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"/>"
`;
